### ISSUE TITLE
Potential mismatch between Stage1/Stage2 evaluation semantics and final checkpoint selection

### ISSUE THREAD

**Issue by Areuagay** [YetZzzzzz/TMDC#1]
Hi, thank you for releasing the code for TMDC. I am currently trying to reproduce the results under the different modality-availability conditions in Table 1, and I encountered a point regarding the Stage1/Stage2 evaluation protocol that I would appreciate some clarification on.

According to the paper, TMDC consists of two stages:

* the IMD stage learns modality-specific and modality-common representations from **complete multimodal data**;
* the IMC stage handles **incomplete multimodal data** and complements the missing modalities.

The experimental section also states that, for example, condition `A` means that **only the audio modality is available**.

However, in the released implementation, I noticed the following behavior.

### 1. Stage1 always uses complete A+T+V input

In `utils_TMDC.py`, `generate_mask()` sets all three modality masks to 1 when `first_stage=True`:

```python
if first_stage:
    audio_mask = np.array([1])
    text_mask = np.array([1])
    visual_mask = np.array([1])
else:
    audio_mask = np.array([1 if 'a' in test_condition else 0])
    text_mask = np.array([1 if 't' in test_condition else 0])
    visual_mask = np.array([1 if 'v' in test_condition else 0])
```

Therefore, during Stage1, runs with `test_condition=a`, `v`, `av`, etc. are all evaluated using complete A+T+V inputs rather than their requested missing-modality condition.

This seems consistent with the paper's definition of the IMD stage.

### 2. Stage1 and Stage2 final metrics are appended to the same arrays

In `train_TMDC.py`, the metrics from every epoch are stored together:

```python
test_accs.append(test_acc)
test_fscores.append(test_fscore)
```

while:

```python
first_stage = True if epoch < args.stage_epoch else False
```

This means that `test_fscores` / `test_accs` contain:

```text
Stage1: complete-modality evaluation
+
Stage2: condition-specific missing-modality evaluation
```

### 3. Final checkpoint selection uses all epochs

At the end of training, the released code uses:

```python
# MOSI / MOSEI
best_index_test = np.argmax(np.array(test_fscores))

# IEMOCAP
best_index_test = np.argmax(np.array(test_accs))
```

without restricting the search to epochs after `stage_epoch`.

Therefore, if the best Stage1 complete-modality score is higher than every Stage2 score, the reported result for a condition such as `A` can actually correspond to a Stage1 A+T+V evaluation.

In my reproduction, this happens for several MOSI conditions. For example, some `a`, `v`, and `av` runs select the same Stage1 epoch and obtain the same Stage1 score, because their Stage1 masks are all A+T+V. Their actual Stage2 condition-specific scores are substantially different.

### My question

Could you please clarify which checkpoint-selection protocol was used to generate Table 1 in the paper?

Was it:

```python
# current released-code behavior
best_epoch = argmax(metric over Stage1 + Stage2)
```

or:

```python
# condition-specific IMC evaluation
best_epoch = stage_epoch + argmax(metric[stage_epoch:])
```

?

My concern is that, under the first protocol, a score labeled as `A`, `V`, or `A,V` may correspond to a complete-modality Stage1 evaluation rather than the modality combination specified in Table 1.

Since Table 1 reports different values for these missing-modality conditions, I would be very grateful if you could clarify the exact evaluation/checkpoint-selection procedure used to produce the published numbers.

If Stage2-only selection was used for the paper, it may be helpful to update the released code so that final condition-specific results are selected only from IMC-stage epochs.

### Two additional implementation questions

While checking the Stage1-to-Stage2 transition, I also noticed:

```python
models.append(model)
```

is used each epoch. Since this appears to append references to the same mutable model object rather than independent historical snapshots, could you confirm whether `model_a`, `model_t`, and `model_v` are intended to represent models from different historical epochs?

I also noticed:

```python
model.state_dict().update(models_a_para_dict)
```

and the corresponding text/video operations.

As I understand PyTorch behavior, modifying the dictionary returned by `state_dict()` does not itself load those values back into the module in the same way as `load_state_dict()`. Could you please confirm whether the intended behavior here is to restore the best Stage1 modality-specific parameters before entering Stage2?

I may be misunderstanding some details of the intended implementation, so any clarification would be very helpful.

Thank you again for releasing the code and for your time.
